prism-agent/service/server/src/main/scala/io/iohk/atala/iam/authentication/oidc/KeycloakAuthenticator.scala

```diff
@@ -1,28 +1,93 @@
 package io.iohk.atala.iam.authentication.oidc
 
 import io.iohk.atala.agent.walletapi.model.BaseEntity
+import io.iohk.atala.agent.walletapi.model.EntityRole
 import io.iohk.atala.iam.authentication.AuthenticationError
 import io.iohk.atala.iam.authentication.AuthenticationError.AuthenticationMethodNotEnabled
 import io.iohk.atala.iam.authentication.AuthenticationError.InvalidCredentials
 import io.iohk.atala.iam.authentication.AuthenticatorWithAuthZ
 import io.iohk.atala.iam.authentication.Credentials
+import io.iohk.atala.shared.utils.Traverse.*
+import pdi.jwt.JwtCirce
+import pdi.jwt.JwtClaim
+import pdi.jwt.JwtOptions
 import zio.*
+import zio.json.ast.Json
 
 import java.util.UUID
 
-final case class KeycloakEntity(id: UUID, accessToken: Option[String] = None, rpt: Option[String] = None)
-    extends BaseEntity
+final class AccessToken private (token: String, claims: JwtClaim, rolesClaimPath: Seq[String]) {
+
+  override def toString(): String = token
+
+  def isRpt: Either[String, Boolean] =
+    Json.decoder
+      .decodeJson(claims.content)
+      .flatMap(_.asObject.toRight("JWT payload must be a JSON object"))
+      .map(_.contains("authorization"))
+
+  def role: Either[String, EntityRole] =
+    extractRoles
+      .map(_.fold(Nil)(_.distinct).toList)
+      .flatMap {
+        case Nil      => Right(EntityRole.Tenant)
+        case r :: Nil => Right(r)
+        case _        => Left(s"Multiple roles is not supported yet.")
+      }
+
+  /** Return a list of roles that is meaningful to the agent */
+  private def extractRoles: Either[String, Option[Seq[EntityRole]]] =
+    Json.decoder
+      .decodeJson(claims.content)
+      .flatMap { json =>
+        val rolesJson = rolesClaimPath.foldLeft(Option(json)) { case (acc, pathSegment) =>
+          acc.flatMap(_.asObject).flatMap(_.get(pathSegment))
+        }
+        rolesJson match {
+          case None => Right(None)
+          case Some(json) =>
+            json.asArray
+              .toRight("Roles claim is not a JSON array of strings.")
+              .map(_.flatMap(_.asString).flatMap(parseRole))
+              .map(Some(_))
+        }
+      }
+
+  private def parseRole(s: String): Option[EntityRole] = {
+    s match {
+      case "agent-admin"  => Some(EntityRole.Admin)
+      case "agent-tenant" => Some(EntityRole.Tenant)
```

Comment:
probably, it's better to rename the `agent-tenant` to `tenant`.
`agent-tenant` couples the tenant to the agent, which might not be a good idea, but it's my personal opinion.

Reply:
I think both are OK at this stage. I'll stick with `admin` as suggested and eventually we might need to support configurable value to be agnostic of the IAM config.